### DISCUSSION TITLE
Best way to automatically group Python `types-*` dependencies with their corresponding packages?

### DISCUSSION THREAD

**Discussion by rushilsrivastava** [renovatebot/renovate#31136] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

Currently, Renovate treats packages and their corresponding type stubs as separate entities, resulting in two different PRs for each. Is there a straightforward way to configure Renovate to group type stubs (`types-*` packages) with their main packages in a single PR, without having to manually configure each group?

For example, if `setuptools` is being updated, I would like `types-setuptools` to be updated in the same PR.

**Comment by rarkins** [comment]
Do they usually have the fact same versions in sync, or major versions in sync, or are completely unsynced?

**Reply by moimael** [reply]
I'd be interested in this, as we are currently grouping those packages manually, and we have a lot of them.

[PEP561](https://peps.python.org/pep-0561/) explains a bit the logic.
[PEP484](https://peps.python.org/pep-0484/) has a [section](https://peps.python.org/pep-0484/#storing-and-distributing-stub-files) about how they are kept in sync.

Pasting it here:

> Third-party stubs have to be versioned using the lowest version of the source package that is compatible. Example: FooPackage has versions 1.0, 1.1, 1.2, 1.3, 2.0, 2.1, 2.2. There are API changes in versions 1.1, 2.0 and 2.2. The stub file package maintainer is free to release stubs for all versions but at least 1.0, 1.1, 2.0 and 2.2 are needed to enable the end user type check all versions. This is because the user knows that the closest lower or equal version of stubs is compatible. In the provided example, for FooPackage 1.3 the user would choose stubs version 1.1.

**Reply by rarkins** [reply]
I assumed that `types-` was an important and special prefix, but PEP561 has only one hit for `types-` and PEP484 has none. Are they always named that way?

In npm we have something similar for `@types/` where we massage the `depName` to be the same as the sibling package and then they get grouped automatically

**Reply by moimael** [reply]
> I assumed that `types-` was an important and special prefix, but PEP561 has only one hit for `types-` and PEP484 has none. Are they always named that way?
> 
> In npm we have something similar for `@types/` where we massage the `depName` to be the same as the sibling package and then they get grouped automatically

Yeah it's not as formalized as with npm. More details here: https://github.com/python/typing/discussions/1503

Assuming that the corresponding types are named `types-{{depName}}` should work 99% of the time though.

**Comment by rarkins** [comment]
I'm wondering if the following might work. Do you have a reproduction repo to test again?

```json
{
  "packageRules": [{
    "matchDatasources": ["pypi"],
    "matchPackageNames": ["types-*"],
    "overrideDepName": "{{replace 'types-' '' depName}}"
}
```
